# Assignment 5 SPARQL queries

I would like you to create the SPARQL query that will answer each of these questions.  Please submit the queries as a Jupyter notebook with the SPARQL kernel activated.  NO programming is required! Submit to GitHub as usual, WITH THE ANSWERS STILL VISIBLE IN THE NOTEBOOK.   Thanks!


UniProt SPARQL Endpoint:  http://sparql.uniprot.org/sparql

In [4]:
%endpoint https://sparql.uniprot.org/sparql

Endpoint set to: https://sparql.uniprot.org/sparql
Request format: JSON

https://sparql.uniprot.org/sparql I checked the queries in this website 

this is very useful https://sparql.uniprot.org/.well-known/sparql-examples/?offset=0&limit=0

this is essential https://en.wikibooks.org/wiki/SPARQL/Expressions_and_Functions

In [2]:
# Q1: 1 POINT How many protein records are in UniProt?
PREFIX up:<http://purl.uniprot.org/core/> 
SELECT (COUNT(DISTINCT  ?protein) AS ?count)
WHERE{ 
        ?protein a up:Protein .
}

count 360157660 Total: 1, Shown: 1

There are 360157660 protein records in Uniprot.

In [3]:
#Q2: 1 POINT How many Arabidopsis thaliana protein records are in UniProt? 
PREFIX up:<http://purl.uniprot.org/core/> 
PREFIX taxon:<http://purl.uniprot.org/taxonomy/> 
SELECT (COUNT(DISTINCT ?protein) AS ?count)
WHERE{ 
        ?protein a up:Protein .

        ?protein up:organism taxon:3702 .  
}

count 136782 Total: 1, Shown: 1

There are 136782 Arabidopsis thaliana proteins in Uniprot.

In [10]:
#Q3: 1 POINT retrieve pictures of Arabidopsis thaliana from UniProt? 
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX wikibase: <http://wikiba.se/ontology#>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX vg: <http://biohackathon.org/resource/vg#>
PREFIX up: <http://purl.uniprot.org/core/>
PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>
PREFIX uberon: <http://purl.obolibrary.org/obo/uo#>
PREFIX taxon: <http://purl.uniprot.org/taxonomy/>
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX SLM: <https://swisslipids.org/rdf/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX sio: <http://semanticscience.org/resource/>
PREFIX sh: <http://www.w3.org/ns/shacl#>
PREFIX schema: <http://schema.org/>
PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>
PREFIX rh: <http://rdf.rhea-db.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX pubmed: <http://rdf.ncbi.nlm.nih.gov/pubmed/>
PREFIX ps: <http://www.wikidata.org/prop/statement/>
PREFIX pq: <http://www.wikidata.org/prop/qualifier/>
PREFIX patent: <http://data.epo.org/linked-data/def/patent/>
PREFIX p: <http://www.wikidata.org/prop/>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX orthodbGroup: <http://purl.orthodb.org/odbgroup/>
PREFIX orthodb: <http://purl.orthodb.org/>


SELECT ?image
WHERE
{
    taxon:3702  <http://xmlns.com/foaf/0.1/depiction>  ?image .
}

#I have tried to use the example 46 of sparql uniprot but there is not image of arabidopsis 
#in https://semantic.eea.europa.eu/sparql

image https://upload.wikimedia.org/wikipedia/commons/3/39/Arabidopsis.jpg https://upload.wikimedia.org/wikipedia/commons/thumb/6/60/Arabidopsis_thaliana_inflorescencias.jpg/800px-Arabidopsis_thaliana_inflorescencias.jpg Total: 2, Shown: 2

In [4]:
#Q4: 1 POINT:  What is the description of the enzyme activity of UniProt Protein Q9SZZ8 
PREFIX up:<http://purl.uniprot.org/core/> 
PREFIX uniprot:<http://purl.uniprot.org/uniprot/> 
PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#> 
PREFIX label:<http://www.w3.org/2004/02/skos/core#>

SELECT ?description
WHERE {
  uniprot:Q9SZZ8 a up:Protein ;
                   up:enzyme ?enzyme .  
  ?enzyme up:activity ?activity .
  ?activity rdfs:label ?description .
}

description Beta-carotene + 4 reduced ferredoxin [iron-sulfur] cluster + 2 H(+) + 2 O(2) = zeaxanthin + 4 oxidized ferredoxin [iron-sulfur] cluster + 2 H(2)O. Total: 1, Shown: 1

In [2]:
#Q5: 1 POINT:  Retrieve the proteins ids, and date of submission, 
#for proteins that have been added to UniProt this year   (HINT Google for “SPARQL FILTER by date”)
# How the assignment is for january maeby 
#I should use 2022 but there is not time yet for add anything because is still 2021 so I will use 2021

PREFIX up: <http://purl.uniprot.org/core/>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?id ?date

WHERE
{
    ?protein a up:Protein ;
               up:created ?date .
    FILTER (?date >= "2021-01-01"^^xsd:date && ?date <= "2021-12-31"^^xsd:date) . #gYear does not work
    BIND (REPLACE(STR(?protein), "http://purl.uniprot.org/uniprot/", "") AS ?id) .
    
} ORDER BY ?date LIMIT 5

# I uses solution 5 of https://www.w3.org/2009/Talks/0615-qbe/ because gYear does not work

id 
 date A0A7G2MLB1 
 2021-02-10 A0A7G9FM31 
 2021-02-10 A0A7G9FPS4 
 2021-02-10 A0A7G9FQN4 
 2021-02-10 A0A7G9FW33 
 2021-02-10 Total: 5, Shown: 5

In [3]:
#Q6: 1 POINT How  many species are in the UniProt taxonomy?
PREFIX taxon:<http://purl.uniprot.org/taxonomy/> 
PREFIX up:<http://purl.uniprot.org/core/> 

SELECT (COUNT(DISTINCT ?taxon) AS ?count)
WHERE{
  ?taxon a up:Taxon .
  ?taxon up:rank up:Species 
} 

count 2029846 Total: 1, Shown: 1

There are 2029846 species.

In [13]:
#Q7: 2 POINT How many species have at least one protein record? 
#(this might take a long time to execute, so do this one last!)

PREFIX up: <http://purl.uniprot.org/core/>

SELECT (COUNT (DISTINCT ?taxon) AS ?count)

WHERE
{
  ?protein a up:Protein;
           up:organism ?taxon .
  ?taxon a up:Taxon;
          up:rank up:Species .
}

count 1057158 Total: 1, Shown: 1

There are 1057158 species with as least one protein record. 

In [2]:
#Q8: 3 points:  find the AGI codes and gene names for all Arabidopsis thaliana  proteins 
#that have a protein function annotation description that mentions “pattern formation”


PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX up:<http://purl.uniprot.org/core/>
PREFIX taxon:<http://purl.uniprot.org/taxonomy/>
PREFIX skos:<http://www.w3.org/2004/02/skos/core#>
SELECT ?agi ?gene_name
WHERE
{
    ?protein a up:Protein .
    ?protein up:organism taxon:3702 .
    ?protein up:encodedBy ?gene .
    ?gene skos:prefLabel ?gene_name .
    ?gene up:locusName ?agi .
    ?protein up:annotation ?annotation.
    ?annotation a up:Function_Annotation .
    ?annotation rdfs:comment ?text .
    FILTER CONTAINS(?text,"pattern formation") .

}


agi 
 gene_name At3g54220 
 SCR At4g21750 
 ATML1 At1g13980 
 GN At5g40260 
 SWEET8 At1g69670 
 CUL3B At1g63700 
 YDA At2g46710 
 ROPGAP3 At1g26830 
 CUL3A At3g09090 
 DEX1 At4g37650 
 SHR At5g55250 
 IAMT1 At3g02130 
 RPK2 At2g42580 
 TTL3 At1g69270 
 RPK1 At5g02010 
 ROPGEF7 At1g66470 
 RHD6 At5g37800 
 RSL1 Total: 17, Shown: 17

From the MetaNetX metabolic networks for metagenomics database SPARQL Endpoint: https://rdf.metanetx.org/sparql
(this slide deck will make it much easier for you!  https://www.metanetx.org/cgi-bin/mnxget/mnxref/MetaNetX_RDF_schema.pdf)

In [1]:
%endpoint https://rdf.metanetx.org/sparql

Endpoint set to: https://rdf.metanetx.org/sparql

In [2]:
#Q9: 4 POINTS:  what is the MetaNetX Reaction identifier (starts with “mnxr”) for the UniProt Protein uniprotkb:Q18A79
PREFIX mnx: <https://rdf.metanetx.org/schema/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX up: <http://purl.uniprot.org/uniprot/>

SELECT ?pept ?reac_label ?reac
WHERE{
    ?pept mnx:peptXref up:Q18A79 .
    ?cata mnx:pept ?pept .
    ?gpr mnx:cata ?cata ;
         mnx:reac ?reac .
    ?reac rdfs:label ?reac_label .

} LIMIT 1

pept 
 reac_label 
 reac https://rdf.metanetx.org/pept/GLGA_CLOD6 
 mnxr165934 
 https://rdf.metanetx.org/reac/mnxr165934 Total: 1, Shown: 1

FEDERATED QUERY - UniProt and MetaNetX



In [ ]:
%endpoint https://sparql.uniprot.org/sparql #otherwise the 10 does not work

In [5]:
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX wikibase: <http://wikiba.se/ontology#>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX vg: <http://biohackathon.org/resource/vg#>
PREFIX up: <http://purl.uniprot.org/core/>
PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>
PREFIX uberon: <http://purl.obolibrary.org/obo/uo#>
PREFIX taxon: <http://purl.uniprot.org/taxonomy/>
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX SLM: <https://swisslipids.org/rdf/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX sio: <http://semanticscience.org/resource/>
PREFIX sh: <http://www.w3.org/ns/shacl#>
PREFIX schema: <http://schema.org/>
PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>
PREFIX rh: <http://rdf.rhea-db.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX pubmed: <http://rdf.ncbi.nlm.nih.gov/pubmed/>
PREFIX ps: <http://www.wikidata.org/prop/statement/>
PREFIX pq: <http://www.wikidata.org/prop/qualifier/>
PREFIX patent: <http://data.epo.org/linked-data/def/patent/>
PREFIX p: <http://www.wikidata.org/prop/>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX orthodbGroup: <http://purl.orthodb.org/odbgroup/>
PREFIX orthodb: <http://purl.orthodb.org/>
PREFIX orth: <http://purl.org/net/orth#>
PREFIX obo: <http://purl.obolibrary.org/obo/>
PREFIX np: <http://nextprot.org/rdf#>
PREFIX nextprot: <http://nextprot.org/rdf/entry/>
PREFIX mnx: <https://rdf.metanetx.org/schema/>
PREFIX mnet: <https://rdf.metanetx.org/mnet/>
PREFIX mesh: <http://id.nlm.nih.gov/mesh/>
PREFIX lscr: <http://purl.org/lscr#>
PREFIX keywords: <http://purl.uniprot.org/keywords/>
PREFIX insdcschema: <http://ddbj.nig.ac.jp/ontologies/nucleotide/>
PREFIX insdc: <http://identifiers.org/insdc/>
PREFIX identifiers: <http://identifiers.org/>
PREFIX glyconnect: <https://purl.org/glyconnect/>
PREFIX glycan: <http://purl.jp/bio/12/glyco/glycan#>
PREFIX genex: <http://purl.org/genex#>
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
PREFIX faldo: <http://biohackathon.org/resource/faldo#>
PREFIX eunisSpecies: <http://eunis.eea.europa.eu/rdf/species-schema.rdf#>
PREFIX ensembltranscript: <http://rdf.ebi.ac.uk/resource/ensembl.transcript/>
PREFIX ensemblterms: <http://rdf.ebi.ac.uk/terms/ensembl/>
PREFIX ensemblprotein: <http://rdf.ebi.ac.uk/resource/ensembl.protein/>
PREFIX ensemblexon: <http://rdf.ebi.ac.uk/resource/ensembl.exon/>
PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>
PREFIX ec: <http://purl.uniprot.org/enzyme/>
PREFIX dc: <http://purl.org/dc/terms/>
PREFIX cco: <http://rdf.ebi.ac.uk/terms/chembl#>
PREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>
PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>
PREFIX bibo: <http://purl.org/ontology/bibo/>
PREFIX allie: <http://allie.dbcls.jp/>
PREFIX GO: <http://purl.obolibrary.org/obo/GO_>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?proteinRef
WHERE
{
    ?protein a up:Protein .
    ?protein up:organism taxon:272563 .
    ?protein up:mnemonic ?mnemonic .
    ?protein up:classifiedWith ?goTerm .
    ?goTerm rdfs:label ?biotype .
    FILTER CONTAINS(?biotype, "starch synthase")
    bind (substr(str(?protein),33) as ?ac)
    bind (IRI(CONCAT("http://purl.uniprot.org/uniprot/",?ac)) as ?proteinRef)
}

proteinRef http://purl.uniprot.org/uniprot/Q18A79 Total: 1, Shown: 1

Find the identifier:

In [7]:
#Q10: 5 POINTS:  What is the official Gene ID (UniProt calls this a “mnemonic”) 
#and the MetaNetX Reaction identifier (mnxr…..) 
#for the protein that has “Starch synthase” catalytic activity in Clostridium difficile (taxon 272563).

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX wikibase: <http://wikiba.se/ontology#>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX vg: <http://biohackathon.org/resource/vg#>
PREFIX up: <http://purl.uniprot.org/core/>
PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>
PREFIX uberon: <http://purl.obolibrary.org/obo/uo#>
PREFIX taxon: <http://purl.uniprot.org/taxonomy/>
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX SLM: <https://swisslipids.org/rdf/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX sio: <http://semanticscience.org/resource/>
PREFIX sh: <http://www.w3.org/ns/shacl#>
PREFIX schema: <http://schema.org/>
PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>
PREFIX rh: <http://rdf.rhea-db.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX pubmed: <http://rdf.ncbi.nlm.nih.gov/pubmed/>
PREFIX ps: <http://www.wikidata.org/prop/statement/>
PREFIX pq: <http://www.wikidata.org/prop/qualifier/>
PREFIX patent: <http://data.epo.org/linked-data/def/patent/>
PREFIX p: <http://www.wikidata.org/prop/>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX orthodbGroup: <http://purl.orthodb.org/odbgroup/>
PREFIX orthodb: <http://purl.orthodb.org/>
PREFIX orth: <http://purl.org/net/orth#>
PREFIX obo: <http://purl.obolibrary.org/obo/>
PREFIX np: <http://nextprot.org/rdf#>
PREFIX nextprot: <http://nextprot.org/rdf/entry/>
PREFIX mnx: <https://rdf.metanetx.org/schema/>
PREFIX mnet: <https://rdf.metanetx.org/mnet/>
PREFIX mesh: <http://id.nlm.nih.gov/mesh/>
PREFIX lscr: <http://purl.org/lscr#>
PREFIX keywords: <http://purl.uniprot.org/keywords/>
PREFIX insdcschema: <http://ddbj.nig.ac.jp/ontologies/nucleotide/>
PREFIX insdc: <http://identifiers.org/insdc/>
PREFIX identifiers: <http://identifiers.org/>
PREFIX glyconnect: <https://purl.org/glyconnect/>
PREFIX glycan: <http://purl.jp/bio/12/glyco/glycan#>
PREFIX genex: <http://purl.org/genex#>
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
PREFIX faldo: <http://biohackathon.org/resource/faldo#>
PREFIX eunisSpecies: <http://eunis.eea.europa.eu/rdf/species-schema.rdf#>
PREFIX ensembltranscript: <http://rdf.ebi.ac.uk/resource/ensembl.transcript/>
PREFIX ensemblterms: <http://rdf.ebi.ac.uk/terms/ensembl/>
PREFIX ensemblprotein: <http://rdf.ebi.ac.uk/resource/ensembl.protein/>
PREFIX ensemblexon: <http://rdf.ebi.ac.uk/resource/ensembl.exon/>
PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>
PREFIX ec: <http://purl.uniprot.org/enzyme/>
PREFIX dc: <http://purl.org/dc/terms/>
PREFIX cco: <http://rdf.ebi.ac.uk/terms/chembl#>
PREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>
PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>
PREFIX bibo: <http://purl.org/ontology/bibo/>
PREFIX allie: <http://allie.dbcls.jp/>
PREFIX GO: <http://purl.obolibrary.org/obo/GO_>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT ?proteinRef ?mnemonic ?reac_label ?biotype
WHERE
    {
    service <http://sparql.uniprot.org/sparql> {
    ?protein a up:Protein .
    ?protein up:organism taxon:272563 .
    ?protein up:mnemonic ?mnemonic .
    ?protein up:classifiedWith ?goTerm .
    ?goTerm rdfs:label ?biotype .
    FILTER CONTAINS(?biotype, "starch synthase")
    bind (substr(str(?protein),33) as ?ac)
    bind (IRI(CONCAT("http://purl.uniprot.org/uniprot/",?ac)) as ?proteinRef)
  }
  service <https://rdf.metanetx.org/sparql> {
    ?pept mnx:peptXref ?proteinRef .
    ?cata mnx:pept ?pept .
    ?gpr mnx:cata ?cata ;
         mnx:reac ?reac .
    ?reac rdfs:label ?reac_label .
  }
}LIMIT 1


proteinRef 
 mnemonic 
 reac_label 
 biotype http://purl.uniprot.org/uniprot/Q18A79 
 GLGA_CLOD6 
 mnxr165934 
 starch synthase activity Total: 1, Shown: 1